File: 2.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = nullptr, *p = nullptr;
        int carry = 0, digit = 0;

        while (l1 || l2) {
            int v1 = l1 ? l1->val : 0;
            int v2 = l2 ? l2->val : 0;
            digit = (v1 + v2 + carry) % 10;
            carry = (v1 + v2 + carry) / 10;

            if (head == nullptr) {
                head = new ListNode(digit);
                p = head;
            } else {
                p->next = new ListNode(digit);
                p = p->next;
            }
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }

        if (carry) p->next = new ListNode(carry);

        return head;
    }
};

int main() {
    Solution s;
    ListNode* l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
    ListNode* l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
    ListNode* result = s.addTwoNumbers(l1, l2);

    while (result) {
        cout << result->val << " ";
        result = result->next;
    }
    cout << endl;

    return 0;
}

File: 2.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

 function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let carry = 0, sum = 0;
    let prev = new ListNode();
    let curr = prev;

    while (l1 != null || l2 != null) {
        let v1 = l1 === null ? 0 : l1.val;
        let v2 = l2 === null ? 0 : l2.val;
        sum = (v1 + v2 + carry) % 10
        carry = Math.floor((v1 + v2 + carry) / 10)
        curr.next = new ListNode(sum, null);
        curr = curr.next

        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
    }

    if (carry > 0) {
        curr.next = new ListNode(carry, null);
    }

    return prev.next;
};